WIP: add endpoint for validating a query limits struct

cstyan opened the pull request:
As part of the per-request query limits we've talked about having a way to validate, during creation or afterwards, whether a query limits structure is valid relative to a tenants overall limits (whether that be the default limits or their actual per tenant overrides).

I'm still trying to grok how we actually wire up our http endpoints and how I could write a test for that, I kept getting ciricular import errors in `querier/http_test.go` trying to create the query limits structure so I backed all of that out for now.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

cstyan committed:
add endpoint for validating a query limits struct against a tenants
limits

Signed-off-by: Callum Styan <callumstyan@gmail.com>

jeschkies commented on pkg/querier/http.go:
```diff
@@ -438,6 +438,26 @@ func (q *QuerierAPI) IndexStatsHandler(w http.ResponseWriter, r *http.Request) {
+func (q *QuerierAPI) LimitsValidationHandler(w http.ResponseWriter, r *http.Request) {
+	// do we maybe need to extract the header here and inject into the ctx?
+
+	// we shouldn't support multiple tenant ids for this endpoint
+	tenantIDs, err := tenant.TenantIDs(r.Context())
```
Is that so? Why not? Do we need a tenant at all?

jeschkies commented on pkg/querier/http.go:
```diff
@@ -438,6 +438,26 @@ func (q *QuerierAPI) IndexStatsHandler(w http.ResponseWriter, r *http.Request) {
+	err = q.limits.ValidateQueryLimits(r.Context(), tenantIDs[0])
+
+	if err == nil {
+		return
```
Is it automatically HTTP OK?

jeschkies commented on pkg/util/explain/explain.go:
```diff
@@ -0,0 +1,15 @@
+	"strings"
+)
+
+func Pretty(level int, explain string) string {
```
That seems from the last hackathon 😊

jeschkies commented on pkg/util/querylimits/limiter.go:
```diff
@@ -24,6 +25,66 @@ func NewLimiter(log log.Logger, original limiter.CombinedLimits) *Limiter {
+func (l *Limiter) ValidateQueryLimits(ctx context.Context, userID string) error {
+	var isErr bool
+	var err []byte
+	err = append(err, []byte("the following limits from the query limits struct violate your tenants limits; ")...)
```
Ah. Got it.

jeschkies commented on pkg/util/querylimits/limiter.go:
```diff
@@ -24,6 +25,66 @@ func NewLimiter(log log.Logger, original limiter.CombinedLimits) *Limiter {
+	// don't need to check required labels, we union the two sets rather than choosing one or the other
+
+	origRequiredNumberLabels := l.CombinedLimits.RequiredNumberLabels(ctx, userID)
+	if requestLimits.RequiredNumberLabels > origRequiredNumberLabels {
```
I think this is the other way around. A higher number is stricter.

cstyan committed:
fix wiring of the new handler, have all the processing be done in the
querylimits package because otherwise we have to figure out how to deal
with a circular import

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
fix test

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
fix lint

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
another test fix

Signed-off-by: Callum Styan <callumstyan@gmail.com>
